File: problem/F4-Network_test_maxcho/F4-Dockerfiles/src/main.c
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <unistd.h>

int main () {
    char host [100];
    int port;

    // 1. 讀取輸入
    if (scanf ("%99s %d" , host , &port) != 2) return 0;

    printf ("debug: resolving hostname %s\n" , host);
    struct hostent *he = gethostbyname (host);
    if (!he) {
        printf ("fail (debug: dns resolution failed)\n");
        return 0;
    }

    // 2. 建立 Socket
    int sock = socket (AF_INET , SOCK_STREAM , 0);
    struct sockaddr_in server;
    server.sin_family = AF_INET;
    server.sin_port = htons (port);
    server.sin_addr = *((struct in_addr *)he->h_addr_list [0]);
    printf ("debug: IP resolved to %s\n" , inet_ntoa (server.sin_addr));

    if (connect (sock , (struct sockaddr *)&server , sizeof (server)) < 0) {
        printf ("fail (debug: connection failed)\n");
        close (sock);
        return 0;
    }

    // 3. 發送 HTTP Request
    printf ("debug: sending HTTP GET...\n");
    char msg [256];
    snprintf (msg , sizeof (msg) , "GET / HTTP/1.0\r\nHost: %s\r\n\r\n" , host);
    send (sock , msg , strlen (msg) , 0);

    // 4. 接收回應
    printf ("debug: receiving data...\n");
    char buffer [4096];
    int total_len = 0;
    int len;

    while ((len = recv (sock , buffer + total_len , sizeof (buffer) - total_len - 1 , 0)) > 0) {
        total_len += len;
        if (total_len >= sizeof (buffer) - 1) break;
    }
    buffer [total_len] = '\0';

    printf ("debug: total bytes: %d\n" , total_len);

    // 5. 驗證內容
    int is_good = 0;
    if (port == 8000) { // env-python
        if (strstr (buffer , "Hello from Server Container!")) {
            printf ("debug: Matched Python env signature\n");
            is_good = 1;
        }
        else {
            printf ("fail (debug: Python env signature not found)\n");
        }
    }
    else if (port == 8080) { // env-cpp
        if (strstr (buffer , "Hello from C++ File!")) {
            printf ("debug: Matched C++ env signature\n");
            is_good = 1;
        }
        else {
            printf ("fail (debug: C++ env signature not found)\n");
        }
    }
    else {
        printf ("fail (debug: Unknown port %d)\n" , port);
    }

    if (is_good) {
        printf ("good\n");
    }

    close (sock);
    return 0;
}

File: problem/F4-Network_test_maxcho/F4-Dockerfiles/src/main.py
import socket
import sys


def test_custom_envs():
    host = "localhost"
    port = 80

    # 1. 讀取輸入
    try:
        line = sys.stdin.read().strip()
        if line:
            parts = line.split()
            if len(parts) >= 2:
                host = parts[0]
                port = int(parts[1])
    except Exception:
        pass

    print(f"debug: Target is {host}:{port}")

    try:
        # 2. 建立連線
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5)

        print("debug: Connecting...")
        sock.connect((host, port))
        print("debug: Connected!")

        # 3. 發送 HTTP Request
        print("debug: Sending GET request...")
        request = f"GET / HTTP/1.0\r\nHost: {host}\r\n\r\n"
        sock.sendall(request.encode())

        # 4. 接收回應
        print("debug: Receiving response...")
        response = b""
        while True:
            chunk = sock.recv(4096)
            if not chunk:
                break
            response += chunk

        response_str = response.decode("utf-8", errors="replace")
        print(f"debug: Response length: {len(response_str)}")
        print(f"debug: Content snippet: {repr(response_str[:100])}...")

        # 5. 驗證內容 (針對 v1 的兩個服務)
        is_good = False

        if port == 8000:  # env-python
            if "Hello from Server Container!" in response_str:
                print("debug: Matched Python env signature")
                is_good = True
            else:
                print("fail (debug: Python env signature not found)")

        elif port == 8080:  # env-cpp
            if "Hello from C++ File!" in response_str:
                print("debug: Matched C++ env signature")
                is_good = True
            else:
                print("fail (debug: C++ env signature not found)")
        else:
            print(f"fail (debug: Unknown port {port})")

        if is_good:
            print("good")

        sock.close()
    except Exception as e:
        print(f"fail (debug: Exception {e})")


if __name__ == "__main__":
    test_custom_envs()
